pymeasure/instruments/keithley/keithley6487.py

```diff
@@ -0,0 +1,395 @@
+        self.write(":SOUR:VOLT:STAT ON")
+
+    def disable_source(self):
+        """Disables the source voltage of the instrument"""
+        self.write(":SOUR:VOLT:STAT OFF")
+
+    def measure_current(self, nplc=1, current=1.05e-4, auto_range=True):
+        """ Configures the measurement of current.
+
+        :param nplc: Number of power line cycles (NPLC) from 0.01 to 10
+        :param current: Upper limit of current in Amps, from -0.021 A to 0.021 A
+        :param auto_range: Enables auto_range if True, else uses the set current
+        """
+        log.info("%s is measuring current." % self.name)
+        self.write(":SENS:FUNC:CURR;"
+                   ":SENS:CURR:NPLC %f;:FORM:ELEM READ;" % nplc)
+        if auto_range:
+            self.write(":SENS:CURR:RANG:AUTO 1;")
+        else:
+            self.current_range = current
+        self.check_errors()
+
+    def auto_range_current(self):
+        """ Configures the current measurement to use an automatic range.
+        """
+        self.write(":SENS:CURR:RANG:AUTO 1")
+
+    display_enabled = Instrument.control(
+        ":DISP:ENAB?", ":DISP:ENAB %d",
+        """ A boolean property that controls whether or not the display of the
+        sourcemeter is enabled. Valid values are True and False. """,
+        values={True: 1, False: 0},
+        map_values=True,
+    )
+
+    @property
+    def error(self):
+        """ Returns a tuple of an error code and message from a
+        single error. """
+        err = self.values(":system:error?")
+        if len(err) < 2:
+            err = self.read()  # Try reading again
+        code = err[0]
+        message = err[1].replace('""', '')
+        return (code, message)
+
+    def check_errors(self):
+        """ Logs any system errors reported by the instrument.
+        """
+        code, message = self.error
+        while code != 0:
+            t = time.time()
+            log.info("Keithley 6487 reported error: %s, %s" % (code, message))
+            code, message = self.error
+            if (time.time() - t) > 10:
+                log.warning("Timed out for Keithley 6487 error retrieval.")
+
+    def reset(self):
+        """ Resets the instrument and clears the queue.  """
+        self.write("status:queue:clear;*RST;:stat:pres;:*CLS;")
+
+    def ramp_to_voltage(self, target_voltage, steps=30, pause=20e-3):
+        """ Ramps to a target voltage from the set voltage value over
+        a certain number of linear steps, each separated by a pause duration.
+
+        :param target_voltage: A voltage in Volts
+        :param steps: An integer number of steps
+        :param pause: A pause duration in seconds to wait between steps
+        """
+        voltages = np.linspace(
+            self.source_voltage,
+            target_voltage,
+            steps
+        )
+        for voltage in voltages:
+            self.source_voltage = voltage
+            time.sleep(pause)
```

Comment:
Timing in python cannot be very accurate. Ramp shape can be distorted.

Reply:
After a quick read in the manual, this could/should be replaced with a sweep in the instrument itself, `SOUR:VOLT:SWE`
See table 6.3 on page 106 of [reference manual](https://download.tek.com/manual/6487-901-01(B-Mar2011)(Ref).pdf)